iceoryx_hoofs/source/posix_wrapper/periodic_timer.cpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2021 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_hoofs/posix_wrapper/periodic_timer.hpp"
+#include "iceoryx_hoofs/posix_wrapper/posix_call.hpp"
+
+using namespace iox::units::duration_literals;
+
+namespace iox
+{
+namespace posix
+{
+PeriodicTimer::PeriodicTimer(const iox::units::Duration interval) noexcept
+    : m_interval(interval)
+{
+    auto timerSemaphore = posix::Semaphore::create(posix::CreateUnnamedSharedMemorySemaphore, 0U);
+    cxx::Ensures(!timerSemaphore.has_error() && "Could not create Semaphore for PeriodicTimer!");
+    m_waitSemaphore = std::move(timerSemaphore.value());
+    start();
+}
+
+void PeriodicTimer::start() noexcept
+{
+    stop();
+    auto waitResult = m_waitSemaphore.timedWait(m_interval);
+    cxx::Ensures(!waitResult.has_error());
+    m_timeForNextActivation = now().value() + m_interval;
+}
+
+void PeriodicTimer::start(const iox::units::Duration interval) noexcept
+{
+    m_interval = interval;
+    start();
+}
+
+void PeriodicTimer::stop() noexcept
+{
+    if (*(m_waitSemaphore.getValue()) == static_cast<int>(posix::SemaphoreWaitState::TIMEOUT))
+    {
+        auto stopResult = m_waitSemaphore.post();
+        cxx::Ensures(!stopResult.has_error());
+    }
+}
+
+cxx::expected<units::Duration, TimerErrorCause> PeriodicTimer::now() noexcept
+{
+    struct timespec ts;
+    auto result =
+        posixCall(clock_gettime)(CLOCK_REALTIME, &ts).failureReturnValue(-1).evaluate();
+
+    if (result.has_error())
+    {
+        return createErrorCodeFromErrNo(result.get_error().errnum);
+    }
+
+    return cxx::success<units::Duration>(ts);
+}
+
+cxx::expected<iox::posix::WaitResult, TimerErrorCause> PeriodicTimer::wait(TimerCatchupPolicy policy) noexcept
+{
+    // To check if the TIMER is active (if the sempahore is acquired)
+    if (*(m_waitSemaphore.getValue()) == static_cast<int>(posix::SemaphoreWaitState::TIMEOUT))
+    {
+        if (now().value() > m_timeForNextActivation)
+        {
+            if (policy == TimerCatchupPolicy::IMMEDIATE_TICK)
+            {
+                m_timeForNextActivation = now().value();
```

Comment:
Again, accessing a value without checking it if its valid. Please use the `and_then`/`or_else` pattern wherever possible to avoid such mistakes.

Reply:
Yes. modified to check if the expected has errors.